packages/core/src/application_init.ts

```diff
@@ -27,6 +27,56 @@ import {noop} from './util/noop';
  *
  * @see `ApplicationInitStatus`
  *
+ * @usageNotes
+ *
+ *
+ * The following module provides a function returning a promise to the `APP_INITIALIZER` token.
+ *
+ * ```
+ *  function initializeApp(): Promise<any> {
+ *    return new Promise((resolve, reject) => {
+ *      // Do some asynchronous stuff
+ *      resolve();
+ *    });
+ *  }
+ *
+ *  @NgModule({
+ *   imports: [BrowserModule],
+ *   declarations: [AppComponent],
+ *   bootstrap: [AppComponent],
+ *   providers: [{
+ *     provide: APP_INITIALIZER,
+ *     useFactory: () => initializeApp,
+ *     multi: true
+ *    }]
+ *   })
+ *  export class AppModule {}
+ * ```
+ *
+ * It is also possible to provide a function returning an observable to the `APP_INITIALIZER` token.
```

Comment:
I'd propose re-phrasing this a bit too and adding a note on `HttpClient` usage (that it's a demo-only thing, not mandatory for an initializer function to work with observables).
```suggestion
 * It's also possible to configure a multi-provider using `APP_INITIALIZER` token and a function returning an observable, see an example below. Note: the `HttpClient` in this example is used for demo purposes to illustrate how the factory function can work with other providers available through DI.
```